neqo-transport/src/connection/tests/zerortt.rs

```diff
@@ -341,3 +341,89 @@ fn zero_rtt_loss_accepted() {
+///
+/// Previously Neqo client would prematurely declare 0-RTT stream data frame as
+/// lost based on time threshold.
+///
+/// See <https://github.com/mozilla/neqo/issues/2476> for details.
+#[test]
+fn zero_rtt_no_time_threshold_loss_detection_before_first_ack() {
+    const RTT: Duration = Duration::from_millis(4);
+    let mut now = now();
+    let mut client = default_client();
+    let mut server = default_server();
+    connect(&mut client, &mut server);
+    let token = exchange_ticket(&mut client, &mut server, now);
+
+    let mut client = default_client();
+    client
+        .enable_resumption(now, token)
+        .expect("should set token");
+    let mut server = resumed_server(&client);
+
+    // Send ClientHello.
+    let client_hs = client.process_output(now);
+    let client_hs2 = client.process_output(now);
+    assert!(client_hs.as_dgram_ref().is_some() && client_hs2.as_dgram_ref().is_some());
+
+    // Attempt sending 0-RTT stream data.
+    let client_stream_id = client.stream_create(StreamType::UniDi).unwrap();
+    client.stream_send(client_stream_id, &[1, 2, 3]).unwrap();
+    // Blocked due to pacing.
+    let delay = client.process_output(now).callback();
+    // Wait for pacer, then send.
+    let client_0rtt = client.process_output(now + delay);
+    assert!(client_0rtt.as_dgram_ref().is_some());
+
+    now += RTT / 2;
+
+    // Receive ClientHello.
+    server.process_input(client_hs.dgram().unwrap(), now);
+    let server_hs = server.process(client_hs2.dgram(), now);
+    assert!(server_hs.as_dgram_ref().is_some()); // ServerHello, etc...
+
+    now += delay;
+
+    // Receive 0-RTT stream data.
+    let server_process_0rtt = server.process(client_0rtt.dgram(), now);
+    assert!(server_process_0rtt.as_dgram_ref().is_some());
+
+    let server_stream_id = server
+        .events()
+        .find_map(|evt| match evt {
+            ConnectionEvent::NewStream { stream_id, .. } => Some(stream_id),
+            _ => None,
+        })
+        .expect("should have received a new stream event");
+    assert_eq!(client_stream_id, server_stream_id.as_u64());
+
+    now += RTT / 2;
+
+    // Process Server packets. Note that this will not contain an ACK for the
+    // 0-RTT stream data, given that the server does not yet have 1-RTT keys and
+    // it will delay the ACK.
+    let _client_pkt1 = client.process(server_hs.dgram(), now);
+    let _client_pkt2 = client.process(server_process_0rtt.dgram(), now);
+    assert!(matches!(
+        client.process_output(dbg!(now)),
+        Output::Callback(_)
+    ));
+
+    now += RTT / 2;
+
+    // Despite the missing ACK from the server, expect client to neither declare
+    // 0-RTT stream data lost, nor attempt to retransmit it. Given that thus far
+    // no Application space ACK has been received, time threshold based loss
+    // detection does not yet apply.
+    assert!(client.process_output(dbg!(now)).dgram().is_none());
```

Comment:
```suggestion
    assert!(client.process_output(now).dgram().is_none());
```

Are you sure that the amount of elapsed time is more than the time loss recovery threshold?  I think that you are at a bit more than 1.5RTT at this point (plus the pacing delay).  That's more than 9/8 RTT, so I guess that's OK.  If my logic is correct, maybe just say that.

Reply:
> Are you sure that the amount of elapsed time is more than the time loss recovery threshold?

Yes, the elapsed time is more than the time loss recovery threshold. Executing this test without the fix in this pull request, the `client` does falsely declare the 0-RTT application data as lost in `client.process_output(dbg!(now)).dgram().is_none()` above.

More specifically it prints the following log lines:

```
0.023 TRACE detect lost ap: now=Instant { tv_sec: 20664, tv_nsec: 609522841 } delay=10.125ms
0.023 TRACE lost=0, time sent Instant { tv_sec: 20664, tv_nsec: 598522841 } is before lost_delay 10.125ms
```

Executing the following `if` arm:

https://github.com/mozilla/neqo/blob/ddb88ac993330379335d921eb68dbd6211fe2840/neqo-transport/src/recovery/mod.rs#L336-L340

>  That's more than 9/8 RTT, so I guess that's OK. If my logic is correct, maybe just say that.

Good idea. Will add a comment above. Thanks for the thorough review @martinthomson.

Reply:
For the record, done in https://github.com/mozilla/neqo/pull/2492/commits/30cdfaefd1ba3ceb7b81e8bfabdce005f2108054.